Career stats, new collections trackers, and general fixes/cleanup

SethFusion opened the pull request:
I made quite a few changes, the biggest of which (from a project vision perspective) is adding smods as a dependency, but I don't want to step on your toes or commandeer your project. So feel free to remove whatever you don't like. Also, I'm sure this could use some cleaning up and testing. I will probably keep making some changes on my end, there are still a few more trackers I want to add. Let me know if there are any specific dos or don'ts, or anything you have been scratching your head over. Maybe I can take a look. :)

SethFusion committed:
Fix Space Joker lovely patch to work with SMODS BETA-1016c

SethFusion committed:
Added jokers + fixed check_record to only apply when you own the Joker where needed

SethFusion committed:
Attempted fix at display/print inconsistannt data

SethFusion committed:
added WIP Career Stats page and API method

SethFusion committed:
removed non-smods patch

SethFusion committed:
added smods and lovely as dependencies

SethFusion committed:
removed code used as reference

SethFusion committed:
reformatted career stats to have two columns

SethFusion committed:
added additional tracked records

SethFusion committed:
renamed some locale variables for consistancy

SethFusion committed:
Added tag and blind usage trackers

SethFusion committed:
various fixes to UI weirdness

SethFusion committed:
fixed win rate logic

SethFusion committed:
Added playing card tracker

SethFusion committed:
renamed method for code consistency

SethFusion committed:
added total counts and blind type selector

SethFusion committed:
fixed stencil issue

SethFusion committed:
added option to hide records during run

SethFusion committed:
Merge pull request #1 from SethFusion/new_stats_testing

New stats testing

SethFusion committed:
last minute fix :(

GauntletGames-2086 commented:
My thoughts on this before I do any deep diving on the changes:
- This mod will not be dependent on SMODS. I made this decision from the start because nothing in the mod needed to interacted with SMODS API directly (at worst I had to copy some code over) and keeping parity isn't too difficult.
- Career stats are a great addition.
- Not a fan of adding Tags, Blinds, or playing cards into the "Card Stats" menu and renaming it "Collection Stats". This is confusing as playing cards aren't part of the "Collection" menu itself, and with the rest of the collections menu to add this menu will likely get so big that it overflows and goes offscreen. Tags also waste the menu given that we only have one stat tracked ("Times Redeemed"). I'd rather keep this as "Card Stats" and restrict it to such.

Note that I will be reworking the "Stats" menu itself to be tabbed, so don't worry too much about if the stats splitting will take too much space.

SethFusion committed:
Removed SMODS dependencies

SethFusion commented:
My life got very busy for a little while, but I got some time so I might come back to this....

- Totally understandable. I _think_ I removed all of the smods specific things I added/removed. Probably could use some more testing. 
- Thank you! I think there is a lot of room for 'high score' type stats instead of trackers. I added quite a few myself but there are probably other things I haven't thought of. I think it also adds something interesting for other mods to call FlowerPot as a dependency, so I wanted to make it as simple as possible to get a value stored and displayed. 
- I understand the hesitation here, and I went back and forth for a while on what to name it. 'Collection' just seemed like the best approximation for the information captured. I still like seeing some of the information being tracked for blinds and tags and whatnot, so my question to you is: Would you like it to go somewhere else or just remove it all completely? Ultimately it's your decision. 
- If you want to keep some of it, I disagree that tags waste the menu because they only track one stat. The purpose of a framework is to provide a foundation to be built off. With tags having a dedicated tab, another mod could add more stats to track (should they provide some functionality that is worth tracking). Consumables also track one stat: times used, and to me they fall under the same category. Just food for thought.

GauntletGames-2086 commented:
> My life got very busy for a little while, but I got some time so I might come back to this....

Neither do it xdd, I would've done the edits and merged already if that was the case.

> * I understand the hesitation here, and I went back and forth for a while on what to name it. 'Collection' just seemed like the best approximation for the information captured. I still like seeing some of the information being tracked for blinds and tags and whatnot, so my question to you is: Would you like it to go somewhere else or just remove it all completely? Ultimately it's your decision.

I do think that Blinds and Tags should have their own buttons on the menu, however I don't think we can just do that atm. The current Stats page UI sucks given that at some point it's going to cause some UI placement issues (so many buttons that it pushes the "Back" button offscreen, adding a 3rd column could make the UI so long that it goes offscreen). My current idea for a solution to that problem is to do a mini-overhaul on the Stats page by giving it tabs that house the separate types of stats. I mentioned it briefly in the last sentence of the PR comment, but honestly it kind of now is a requirement for this PR so it feels necessary to clarify what I meant. It also means that if you get back to this before I do then you wouldn't have to wait for me to add it. 

I'm imagining something like this:
- Tab one is "Progress" with just the UI parts from the vanilla Stats tab, minus card stats. I can turn this later into something more dedicated to tracking custom Stake wins but that's out-of-scope for this PR.
- Tab two is "Career Stats", where the UI for the Career Stats part of this PR is present.
- Third tab is "Collection", which houses all of the buttons we currently have to display collection stats (Cards, Poker Hands, Decks) along with both the Blinds and Tags button. Any future additions made that display stats can be put here. 

I believe this should cover all the bases we need atm, but I could be forgetting something.

> * If you want to keep some of it, I disagree that tags waste the menu because they only track one stat. The purpose of a framework is to provide a foundation to be built off. With tags having a dedicated tab, another mod could add more stats to track (should they provide some functionality that is worth tracking). Consumables also track one stat: times used, and to me they fall under the same category. Just food for thought.

In hindsight, having the menus exist even if it is for just one stat is entirely fine. Keep it.
